src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,275 @@
+        if (!sigResult)
+        {
+            JLOG(ctx.j.warn()) << "Batch: invalid batch txn signature.";
+            return temBAD_SIGNATURE;
+        }
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+Batch::preclaim(PreclaimContext const& ctx)
+{
+    if (!ctx.view.rules().enabled(featureBatch))
+        return temDISABLED;
+
+    return tesSUCCESS;
+}
+
+TER
+Batch::doApply()
+{
+    Sandbox sb(&ctx_.view());
+    bool changed = false;
+    auto const flags = ctx_.tx.getFlags();
+
+    AccountID const outerAccount = ctx_.tx.getAccountID(sfAccount);
+
+    TER result = tesSUCCESS;
+    ApplyViewImpl& avi = dynamic_cast<ApplyViewImpl&>(ctx_.view());
+    OpenView subView(&sb);
+
+    auto const& txns = ctx_.tx.getFieldArray(sfRawTransactions);
+    bool const not3rdParty = std::any_of(
+        txns.begin(), txns.end(), [outerAccount](STObject const& txn) {
+            return txn.getAccountID(sfAccount) == outerAccount;
+        });
+
+    for (STObject txn : txns)
+    {
+        OpenView innerView(&subView);
+
+        STTx const stx = STTx{std::move(txn)};
+        auto const [ter, applied] =
+            ripple::apply(ctx_.app, innerView, stx, tapFAIL_HARD, ctx_.journal);
+
+        if (applied)
+            innerView.apply(subView);
+
+        changed = true;
+
+        // Add Inner Txn Metadata
+        STObject meta{sfBatchExecution};
+        std::string res = transToken(ter);
+        meta.setFieldVL(sfInnerResult, ripple::Slice{res.data(), res.size()});
+        meta.setFieldU16(sfTransactionType, stx.getTxnType());
+        if (ter == tesSUCCESS || isTecClaim(ter))
+            meta.setFieldH256(sfTransactionHash, stx.getTransactionID());
+        avi.addBatchExecutionMetaData(std::move(meta));
+
+        if (ter != tesSUCCESS)
+        {
+            // Atomic Revert on non tec failure
+            if (!isTecClaim(ter))
+            {
+                JLOG(ctx_.journal.warn()) << "Batch: Inner txn failed." << ter;
+                result = tecBATCH_FAILURE;
+                changed = false;
+                break;
+            }
+
+            if (flags & tfUntilFailure)
+            {
+                result = tesSUCCESS;
+                break;
+            }
+            if (flags & tfOnlyOne)
+            {
+                continue;
+            }
+            if (flags & tfAllOrNothing)
+            {
+                result = tecBATCH_FAILURE;
+                changed = false;
+                break;
+            }
+        }
+
+        if (ter == tesSUCCESS && flags & tfOnlyOne)
+        {
+            result = tesSUCCESS;
+            break;
+        }
+    }
+
+    // Apply SubView & PreviousFields
+    if (changed)
+    {
+        // Only required when not 3rd Party
+        if (not3rdParty)
+            ctx_.batchPrevious(avi);
+
+        ctx_.applyOpenView(subView);
+    }
+
+    sb.apply(ctx_.rawView());
+    return result;
+}
+
+XRPAmount
+Batch::calculateBaseFee(ReadView const& view, STTx const& tx)
```

Comment:
nit: this should be at the top (to match how it exists in other transaction cpp files).